lib/topological_inventory/schema/default.rb

```diff
@@ -123,6 +124,102 @@ def add_tagging_collection(model, manager_ref: [:source_ref])
       end
 
+      def add_service_instance_tasks
+        add_collection(:service_instance_tasks, inventory_collection_builder, {}, {:without_model_class => true}) do |builder|
+          add_default_properties(builder)
+          add_default_values(builder)
+
+          builder.add_dependency_attributes(
+            :service_instances => ->(persister) { [persister.collections[:service_instances]] }
+          )
+
+          save_block = lambda do |source, tasks_collection|
+            service_instance_tasks_custom_save(source, tasks_collection)
+          end
+
+          builder.add_properties(:custom_save_block => save_block)
+        end
+      end
+
+      def service_instance_tasks_custom_save(source, tasks_collection)
+        service_instance_collection = tasks_collection.dependency_attributes[:service_instances]&.first
+        src_refs = service_instance_collection&.data.to_a.collect { |inventory_object| inventory_object.source_ref }
+        return if src_refs.blank?
+
+        # Get running tasks
+        tasks_values = Task.where(:state => 'running', :target_type => 'ServiceInstance', :source_id => source.id)
+                             .pluck(:id, :target_source_ref)
+        tasks_id, tasks_source_ref = [], []
+        tasks_values.each do |attrs|
+          tasks_id << attrs[0]
+          tasks_source_ref << attrs[1]
+        end
+
+        # Load saved service instances (IDs needed)
+        svc_instances_values = ServiceInstance.where(:source_ref => tasks_source_ref).pluck(:id, :external_url, :source_ref, Arel.sql("extra->'finished'"), Arel.sql("extra->'status'"))
+        return if svc_instances_values.blank?
+
+        # Updating Tasks
+        # service_instance_tasks_update_effective(svc_instances_values)
+        service_instance_tasks_update_ineffective(svc_instances_values, tasks_id, tasks_source_ref)
+      end
+
+      def task_update_values(svc_instance_id, external_url, status, finished_timestamp)
+        {
+          :state  => finished_timestamp.blank? ? 'running' : 'completed',
+          :status => %w[error failed].include?(status) ? 'error' : 'ok', # TODO: ansible-tower specific, normalize in collector
+          :context => {
+            :remote_status => status,
+            :service_instance => {
+              :id => svc_instance_id,
+              :url => external_url
+            }
+          }
+        }
+      end
+
+      # This method is updating one by one using ActiveRecord
+      def service_instance_tasks_update_ineffective(service_instances_values, tasks_id, tasks_source_ref)
+        tasks = {}
+        service_instances_values.each do |attrs|
+          id, external_url, source_ref, finished_timestamp, status = attrs[0], attrs[1], attrs[2], attrs[3], attrs[4]
+
+          task_id = tasks_id[tasks_source_ref.index(source_ref)]
+          tasks[task_id] = task_update_values(id, external_url, status, finished_timestamp)
+
+          Task.update(tasks.keys, tasks.values)
+        end
```

Comment:
I think this is more ineffective than it has to be:

```ruby
# Set up 1000 running tasks and service instances
1_000.times { source.service_instances.create!(:tenant => t, :source => s, :source_ref => SecureRandom.uuid) }
source.service_instances.each { |si| Task.create!(:state => "running", :status => "ok", :target_type => "ServiceInstance", :source => s, :target_source_ref => si.source_ref, :tenant => t) }

tasks_by_source_ref = Task.where(:state => 'running', :target_type => 'ServiceInstance', :source_id => source.id, :target_source_ref => service_instances.pluck(:source_ref)).index_by(&:target_source_ref)
service_instances.find_in_batches { |group| ActiveRecord::Base.transaction { group.each { |service_instance| tasks_by_source_ref[service_instance.source_ref].update(:context => {:service_instance => {:id => ser
vice_instance.id}}) } } }
```

Takes `0.70144`

The way you have here:
```
service_instances_values.each do |attrs|
  id, external_url, source_ref, finished_timestamp, status = attrs[0], attrs[1], attrs[2], attrs[3], attrs[4]
  task_id = tasks_id[tasks_source_ref.index(source_ref)]
  tasks[task_id] = {:context => {:service_instance => {:id => id}}}
  end
Task.update(tasks.keys, tasks.values)
```

Takes `4.05301` seconds

Reply:
yeah, my mistake with the update in the loop...

ok, so let's waste time with benchmarking:

**Your way**
`@real=1.7165850750170648, @cstime=0.0, @cutime=0.0, @stime=0.08524299999999996, @utime=1.2653870000000005, @total=1.3506300000000004`

**Current way**
`@real=6.16914131003432, @cstime=0.0, @cutime=0.0, @stime=0.28006200000000003, @utime=2.8168819999999997, @total=3.0969439999999997`

**Current way with transaction**
`@real=2.2340267669642344, @cstime=0.0, @cutime=0.0, @stime=0.10396899999999976, @utime=1.7059960000000007, @total=1.8099650000000005`

**Raw sql**
`@real=0.031836505979299545, @cstime=0.0, @cutime=0.0, @stime=0.0009470000000000312, @utime=0.0013389999999979807, @total=0.002285999999998012`

- raw SQL query is **much much better** than any other solution
- current solution encapsulated to transaction is not as better as yours, but not significantly

If you still don't want raw SQL, can I just use this one with transaction? I don't want to spent another time with rewriting, testing and solving conflicts with rebasing in 2nd task

Reply:
> I don't want to spend another time with rewriting, testing and solving conflicts with rebasing in 2nd task

If you don't have time for this I'll take over this PR for you

I'd rather not just wrap what you have in a transaction and call it done because the whole point is to make this simple and easy to read, what you have here isn't simple and easy to read

Reply:
okay, I don't get it, we discussed that it's not sure that raw sql is faster without benchmarking. Now benchmarking proved it's more than 50x faster and still it isn't good :D 
In your code, it will increase persister's IC saving by 0.5+ seconds... :thinking: 

This PR is not logical for me...

But OK, I'll rewrite it again

Reply:
I think you're missing the point a bit, no one ever said that raw sql wasn't faster but that is a bit like taking a formula 1 car to go to the store.

1000 running tasks in each batch of 1000 service instances isn't realistic, but like we've said all along if it turns out that this is significantly increasing the refresh performance overall then we can optimize it.

Do everything the simple and obvious way first, then measure and optimize what needs to be optimized.

While this is slower it might account for 0.01% of the total refresh time overall (for all collections not just this one).  If it ends up accounting for more like 10% then absolutely we should optimize it.